src/abstractdataframe/join.jl

```diff
@@ -91,6 +91,231 @@ _rename_cols(old_names::AbstractVector{Symbol},
            (renamecols isa Function ? Symbol(renamecols(string(n))) : Symbol(n, renamecols))
            for n in old_names]
 
+prepare_on_col() = throw(ArgumentError("at least one on column required when joining"))
+prepare_on_col(c::AbstractVector) = c
+prepare_on_col(cs::AbstractVector...) = tuple.(cs...)
```

Comment:
There is a function barrier, as joining algorithms are implemented in separate functions. The point is that this way I have one column to join on. There are two benefits:
1. simpler code;
2. one record of data is in a continuous block of memory.

Another approach would be to pass-around all original columns and create these tuples dynamically - as we need to store them as keys in the lookup dictionary.

Reply:
So this is the last point of the PR I think we need to discuss. The reason is:
```
julia> x = [1,2, missing]
3-element Array{Union{Missing, Int64},1}:
 1
 2
  missing

julia> y = [1, missing, 3]
3-element Array{Union{Missing, Int64},1}:
 1
  missing
 3

julia> tuple.(x, y)
3-element Array{Tuple{Any,Any},1}:
 (1, 1)
 (2, missing)
 (missing, 3)
```
which means that if we join on multiple columns that do not have a concrete `eltype` things will be inefficient. It would be good to discuss what would be the best strategy here (and if we can leave it for later).

What currently comes to my mind is:
1. create a custom `AbstractVector` type
2. this `AbstractVector` would be similar to `StructArray`, the only difference is that when indexed into, it would return a lightweight wrapper signaling source and row number.
3. Then for this wrapper we would need to define `isequal`, `isless` and `hash`.

But maybe there is a simpler way to do it?

Reply:
Ah indeed. And why not have `OnCol <: AbstractVector{<:Tuple}`, creating tuples on the fly from `getindex` based on the requested index? Wouldn't that be enough to support `isequal` and `hash` efficiently?

Reply:
One possible advantage of having `OnColRow` is that it would allow computing hashes by processing one column at a time when constructing `OnCol`, storing the result, and returning it from `hash(::OnColRow)`. As I noted somewhere it's supposed to be faster than doing that when the `Dict` lookup calls `hash`, as it can use SIMD (which isn't possible in the complex loop `Dict` uses). This is what `hashrow_cols!` does. Maybe worth checking.